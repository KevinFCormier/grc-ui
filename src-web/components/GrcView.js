/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import resources from '../../lib/shared/resources'
import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import { updateResourceToolbar, updateActiveFilters } from '../actions/common'
import { Loading, Notification } from 'carbon-components-react'
import { GRC_VIEW_STATE_COOKIE, GRC_FILTER_STATE_COOKIE } from '../../lib/shared/constants'
import GrcCardsModule from './modules/GrcCardsModule'
import GrcToggleModule from './modules/GrcToggleModule'
import { filterPolicies, filterFindings, getAvailableGrcFilters, getSavedGrcState, saveGrcState, replaceGrcState, combineResourceFilters } from '../../lib/client/filter-helper'
import { showResourceToolbar, hideResourceToolbar } from '../../lib/client/resource-helper'
import NoResource from './common/NoResource'
import createDocLink from './common/CreateDocLink'
import ResourceFilterBar from './common/ResourceFilterBar'
import msgs from '../../nls/platform.properties'
import _ from 'lodash'
import queryString from 'query-string'
import config from '../../lib/shared/config'

resources(() => {
  require('../../scss/grc-view.scss')
})

class GrcView extends React.Component {

  constructor (props) {
    super(props)
    this.state= {
      viewState: getSavedGrcState(GRC_VIEW_STATE_COOKIE)
    }
    this.onUnload = this.onUnload.bind(this)
    window.addEventListener('beforeunload', this.onUnload)
    this.handleCreatePolicy = this.handleCreatePolicy.bind(this)
    this.updateViewState = this.updateViewState.bind(this)
    this.handleDrillDownClickGrcView = this.handleDrillDownClickGrcView.bind(this)
  }

  componentWillReceiveProps(nextProps) {
    const {refreshControl, grcItems, updateActiveFilters, updateResourceToolbar} = nextProps
    if (!_.isEqual(refreshControl, this.props.refreshControl) ||
        !_.isEqual(grcItems, this.props.grcItems)) {
      const { locale } = this.context
      const displayType = location.pathname.split('/').pop()

      let availableGrcFilters
      switch(displayType) {
      case 'all':
      default:
        availableGrcFilters = getAvailableGrcFilters(grcItems, [], locale)
        break
      case 'findings':
        availableGrcFilters = getAvailableGrcFilters([], grcItems, locale)
        break
      }
      updateResourceToolbar(refreshControl, availableGrcFilters)

      //get (activeFilters ∪ storedFilters) ∩ availableGrcFilters
      const activeFilters = _.cloneDeep(nextProps.activeFilters||{})
      const combinedFilters = combineResourceFilters(activeFilters, getSavedGrcState(GRC_FILTER_STATE_COOKIE), availableGrcFilters)
      //update sessionStorage
      replaceGrcState(GRC_FILTER_STATE_COOKIE, combinedFilters)
      //update active filters
      updateActiveFilters(combinedFilters)
    }
  }

  componentWillUnmount() {
    window.removeEventListener('beforeunload', this.onUnload)
    this.onUnload()
  }

  render() {
    const { locale } = this.context
    const { viewState } = this.state
    const { loading, error, grcItems, activeFilters={}, secondaryHeaderProps, refreshControl, location } = this.props
    hideResourceToolbar()
    if (loading)
      return <Loading withOverlay={false} className='content-spinner' />

    if (error)
      return <Notification title='' className='overview-notification' kind='error'
        subtitle={msgs.get('overview.error.default', locale)} />

    const displayType = location.pathname.split('/').pop()
    let filterGrcItems
    switch(displayType) {
    case 'all':
    default:
      if ((!grcItems || grcItems.length === 0) && !loading) {
        return (
          <NoResource
            title={msgs.get('no-resource.title', [msgs.get('routes.grc', locale)], locale)}
            detail={msgs.get('no-resource.detail.policy', locale)}>
            {createDocLink(locale, this.handleCreatePolicy, msgs.get('routes.create.policy', locale))}
          </NoResource>
        )
      }
      else {
        filterGrcItems = filterPolicies(grcItems, activeFilters, locale)
      }
      break
    case 'findings':
      if ((!grcItems || grcItems.length === 0) && !loading) {
        return (
          <NoResource
            title={msgs.get('no-resource.title', [msgs.get('routes.grc', locale)], locale)}
            detail={msgs.get('no-resource.detail.item', locale)}>
            {createDocLink(locale, this.handleCreatePolicy, msgs.get('routes.create.policy', locale), false)}
          </NoResource>
        )
      }
      else {
        filterGrcItems = filterFindings(grcItems, activeFilters, locale)
      }
      break
    }

    showResourceToolbar()
    const urlParams = queryString.parse(location.search)
    const showGrcCard = urlParams.card==='false' ? false : true
    const grcTabToggleIndex = urlParams.index ? Number(urlParams.index) : 0
    const showGrcTabToggle = urlParams.toggle==='false' ? false : true
    const highLightRowName = urlParams.name ? urlParams.name : ''
    const showSidePanel = urlParams.side==='true' ? true : false
    return (
      <div className='grc-view'>
        <ResourceFilterBar />
        <GrcCardsModule
          displayType={displayType}
          viewState={viewState}
          updateViewState={this.updateViewState}
          grcItems={filterGrcItems}
          activeFilters={activeFilters}
          showGrcCard={showGrcCard}
          handleDrillDownClick={this.handleDrillDownClickGrcView}
        />
        <GrcToggleModule
          displayType={displayType}
          refreshControl={refreshControl}
          grcItems={filterGrcItems}
          secondaryHeaderProps={secondaryHeaderProps}
          locale={locale}
          grcTabToggleIndex={grcTabToggleIndex}
          showGrcTabToggle={showGrcTabToggle}
          highLightRowName={highLightRowName}
          showSidePanel={showSidePanel}
          handleCreatePolicy={this.handleCreatePolicy} />
      </div>
    )
  }

  updateViewState(states) {
    this.setState(prevState=>{
      return {viewState:  Object.assign(prevState.viewState, states)}
    })
  }

  onUnload() {//saved grc view ui setting
    saveGrcState(GRC_VIEW_STATE_COOKIE, this.state.viewState)
  }

  handleDrillDownClickGrcView(key, value, type, level){
    //step 1 add activeFilters when click GrcCardsModule
    //here for severity level, will not update filter here but just update url
    //then acutally update it in componentWillReceiveProps()
    const {updateActiveFilters} = this.props
    const activeFilters = _.cloneDeep(this.props.activeFilters||{})//loadash recursively deep clone
    value = _.startCase(value.replace(' ', '-'))//covert filter name on policy card to start case to match
    let activeSet
    activeFilters[key] ? activeSet = activeFilters[key] : activeSet = activeFilters[key] = new Set()
    activeSet.add(value)
    //add severity level filter
    activeFilters[type] ? activeSet = activeFilters[type] : activeSet = activeFilters[type] = new Set()
    activeSet.add(level)
    replaceGrcState(GRC_FILTER_STATE_COOKIE, activeFilters)
    updateActiveFilters(activeFilters)

    //step 2 update url when click GrcCardsModule
    const paraURL = {}
    paraURL.card=false
    paraURL.toggle=false
    type.toLowerCase()==='cluster' ? paraURL.index=1 : paraURL.index=0
    let urlString = queryString.stringify(paraURL)
    //also append GrcToggleModule search input filter to the end of url if existing
    const curentURL = queryString.parse(location.search)
    if(curentURL.filters && curentURL.filters!==''){
      urlString = `${urlString}&filters=${curentURL.filters}`}
    this.props.history.push(`${this.props.location.pathname}?${urlString}`)
  }

  handleCreatePolicy(){
    this.props.history.push(`${config.contextPath}/policies/create`)
  }
}

GrcView.propTypes = {
  activeFilters: PropTypes.object,
  error: PropTypes.object,
  grcItems: PropTypes.array,
  history: PropTypes.object.isRequired,
  loading: PropTypes.bool,
  location: PropTypes.object,
  refreshControl: PropTypes.object,
  secondaryHeaderProps: PropTypes.object,
  updateActiveFilters: PropTypes.func,
  updateResourceToolbar: PropTypes.func,
}

const mapStateToProps = (state) => {
  const {resourceToolbar: {activeFilters}} = state
  return { activeFilters }
}

const mapDispatchToProps = (dispatch) => {
  return {
    updateResourceToolbar: (refreshControl, availableFilters) => dispatch(updateResourceToolbar(refreshControl, availableFilters)),
    updateActiveFilters: (activeFilters) => dispatch(updateActiveFilters(activeFilters))
  }
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(GrcView))

