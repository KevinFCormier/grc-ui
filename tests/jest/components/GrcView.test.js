/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import { shallow } from 'enzyme'
import { GrcViewItem, GrcViewRefreshControl, GrcViewSecondaryHeaderProps } from './ComponentsTestingData'
import { default as GrcView } from '../../../src-web/components/GrcView'

describe('GrcView view', () => {
  it('renders expand as expected', () => {
    const component = shallow(<GrcView
      loading={!GrcViewItem && false}
      error={null}
      policies={GrcViewItem}
      refreshControl={GrcViewRefreshControl}
      secondaryHeaderProps={GrcViewSecondaryHeaderProps} />)
    expect(component).toMatchSnapshot()
  })
})
