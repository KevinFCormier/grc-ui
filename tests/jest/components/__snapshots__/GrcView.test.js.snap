// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GrcView view renders expand as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <withRouter(Connect(GrcView))
    error={null}
    loading={false}
    policies={
      Array [
        Object {
          "__typename": "Compliance",
          "clusterCompliant": "0/2",
          "metadata": Object {
            "__typename": "Metadata",
            "annotations": Object {
              "policy.mcm.ibm.com/categories": "",
              "policy.mcm.ibm.com/controls": "",
              "policy.mcm.ibm.com/standards": "",
              "seed-generation": "1",
            },
            "name": "my-policy",
            "namespace": "mcm",
            "resourceVersion": "1347399",
            "selfLink": "/apis/policy.mcm.ibm.com/v1alpha1/namespaces/mcm/policies/my-policy",
          },
          "name": "my-policy",
          "namespace": "mcm",
          "placementBindings": Array [
            Object {
              "__typename": "PlacementBinding",
              "metadata": Object {
                "__typename": "Metadata",
                "name": "binding-my-policy",
                "selfLink": "/apis/mcm.ibm.com/v1alpha1/namespaces/mcm/placementbindings/binding-my-policy",
              },
            },
          ],
          "placementPolicies": Array [
            Object {
              "__typename": "PlacementPolicy",
              "metadata": Object {
                "__typename": "Metadata",
                "name": "placement-my-policy",
                "selfLink": "/apis/mcm.ibm.com/v1alpha1/namespaces/mcm/placementpolicies/placement-my-policy",
              },
            },
          ],
          "policyCompliant": "0/2",
          "raw": Object {
            "apiVersion": "policy.mcm.ibm.com/v1alpha1",
            "kind": "Policy",
            "metadata": Object {
              "annotations": Object {
                "policy.mcm.ibm.com/categories": "",
                "policy.mcm.ibm.com/controls": "",
                "policy.mcm.ibm.com/standards": "",
                "seed-generation": "1",
              },
              "creationTimestamp": "2019-07-21T15:53:58Z",
              "finalizers": Array [
                "propagator.finalizer.mcm.ibm.com",
              ],
              "generation": 55,
              "name": "my-policy",
              "namespace": "mcm",
              "resourceVersion": "1347399",
              "selfLink": "/apis/policy.mcm.ibm.com/v1alpha1/namespaces/mcm/policies/my-policy",
              "uid": "c049ec0d-abcf-11e9-8a41-005056a061f1",
            },
            "spec": Object {
              "complianceType": "musthave",
              "namespaces": Object {
                "exclude": Array [
                  "kube-*",
                ],
                "include": Array [
                  "default",
                ],
              },
              "object-templates": Array [
                Object {
                  "complianceType": "musthave",
                  "objectDefinition": Object {
                    "apiVersion": "rbac.authorization.k8s.io/v1",
                    "kind": "RoleBinding",
                    "metadata": Object {
                      "name": "operate-pods-rolebinding",
                    },
                    "roleRef": Object {
                      "apiGroup": "rbac.authorization.k8s.io",
                      "kind": "Role",
                      "name": "operator",
                    },
                    "subjects": Array [
                      Object {
                        "apiGroup": "rbac.authorization.k8s.io",
                        "kind": "User",
                        "name": "admin",
                      },
                    ],
                  },
                  "status": Object {
                    "Validity": Object {},
                  },
                },
              ],
              "remediationAction": "enforce",
            },
            "status": Object {
              "placementBindings": Array [
                "binding-my-policy",
              ],
              "placementPolicies": Array [
                "placement-my-policy",
              ],
              "status": Object {
                "cluster1": Object {
                  "aggregatePoliciesStatus": Object {
                    "my-policy": Object {
                      "compliant": "Compliant",
                    },
                  },
                  "clustername": "cluster1",
                  "compliant": "Compliant",
                },
                "clusterhub": Object {
                  "aggregatePoliciesStatus": Object {
                    "my-policy": Object {
                      "compliant": "Compliant",
                    },
                  },
                  "clustername": "clusterhub",
                  "compliant": "Compliant",
                },
              },
            },
          },
          "remediation": "enforce",
        },
        Object {
          "__typename": "Compliance",
          "clusterCompliant": "1/2",
          "metadata": Object {
            "__typename": "Metadata",
            "annotations": Object {
              "kubectl.kubernetes.io/last-applied-configuration": "{\\"apiVersion\\":\\"policy.mcm.ibm.com/v1alpha1\\",\\"kind\\":\\"Policy\\",\\"metadata\\":{\\"annotations\\":{\\"policy.mcm.ibm.com/categories\\":\\"SystemAndInformationIntegrity\\",\\"policy.mcm.ibm.com/controls\\":\\"MutationAdvisor\\",\\"policy.mcm.ibm.com/standards\\":\\"NIST\\"},\\"name\\":\\"policy-ma\\",\\"namespace\\":\\"mcm\\"},\\"spec\\":{\\"policy-templates\\":[{\\"objectDefinition\\":{\\"apiVersion\\":\\"policies.ibm.com/v1alpha1\\",\\"kind\\":\\"MutationPolicy\\",\\"metadata\\":{\\"label\\":{\\"category\\":\\"System-Integrity\\"},\\"name\\":\\"mutation-policy-example\\"},\\"spec\\":{\\"conditions\\":{\\"ownership\\":[\\"ReplicaSet\\",\\"Deployment\\",\\"DeamonSet\\",\\"ReplicationController\\"]},\\"namespaceSelector\\":{\\"exclude\\":[\\"kube-system\\"],\\"include\\":[\\"default\\",\\"kube-*\\"]},\\"remediationAction\\":\\"enforce\\"}}}],\\"remediationAction\\":\\"enforce\\"}}
",
              "policy.mcm.ibm.com/categories": "SystemAndInformationIntegrity",
              "policy.mcm.ibm.com/controls": "MutationAdvisor",
              "policy.mcm.ibm.com/standards": "NIST",
              "seed-generation": "1",
            },
            "name": "policy-ma",
            "namespace": "mcm",
            "resourceVersion": "1347579",
            "selfLink": "/apis/policy.mcm.ibm.com/v1alpha1/namespaces/mcm/policies/policy-ma",
          },
          "name": "policy-ma",
          "namespace": "mcm",
          "placementBindings": Array [
            Object {
              "__typename": "PlacementBinding",
              "metadata": Object {
                "__typename": "Metadata",
                "name": "binding-ma",
                "selfLink": "/apis/mcm.ibm.com/v1alpha1/namespaces/mcm/placementbindings/binding-ma",
              },
            },
          ],
          "placementPolicies": Array [
            Object {
              "__typename": "PlacementPolicy",
              "metadata": Object {
                "__typename": "Metadata",
                "name": "placement-ma",
                "selfLink": "/apis/mcm.ibm.com/v1alpha1/namespaces/mcm/placementpolicies/placement-ma",
              },
            },
          ],
          "policyCompliant": "1/2",
          "raw": Object {
            "apiVersion": "policy.mcm.ibm.com/v1alpha1",
            "kind": "Policy",
            "metadata": Object {
              "annotations": Object {
                "kubectl.kubernetes.io/last-applied-configuration": "{\\"apiVersion\\":\\"policy.mcm.ibm.com/v1alpha1\\",\\"kind\\":\\"Policy\\",\\"metadata\\":{\\"annotations\\":{\\"policy.mcm.ibm.com/categories\\":\\"SystemAndInformationIntegrity\\",\\"policy.mcm.ibm.com/controls\\":\\"MutationAdvisor\\",\\"policy.mcm.ibm.com/standards\\":\\"NIST\\"},\\"name\\":\\"policy-ma\\",\\"namespace\\":\\"mcm\\"},\\"spec\\":{\\"policy-templates\\":[{\\"objectDefinition\\":{\\"apiVersion\\":\\"policies.ibm.com/v1alpha1\\",\\"kind\\":\\"MutationPolicy\\",\\"metadata\\":{\\"label\\":{\\"category\\":\\"System-Integrity\\"},\\"name\\":\\"mutation-policy-example\\"},\\"spec\\":{\\"conditions\\":{\\"ownership\\":[\\"ReplicaSet\\",\\"Deployment\\",\\"DeamonSet\\",\\"ReplicationController\\"]},\\"namespaceSelector\\":{\\"exclude\\":[\\"kube-system\\"],\\"include\\":[\\"default\\",\\"kube-*\\"]},\\"remediationAction\\":\\"enforce\\"}}}],\\"remediationAction\\":\\"enforce\\"}}
",
                "policy.mcm.ibm.com/categories": "SystemAndInformationIntegrity",
                "policy.mcm.ibm.com/controls": "MutationAdvisor",
                "policy.mcm.ibm.com/standards": "NIST",
                "seed-generation": "1",
              },
              "creationTimestamp": "2019-07-16T13:18:48Z",
              "finalizers": Array [
                "propagator.finalizer.mcm.ibm.com",
              ],
              "generation": 45,
              "name": "policy-ma",
              "namespace": "mcm",
              "resourceVersion": "1347579",
              "selfLink": "/apis/policy.mcm.ibm.com/v1alpha1/namespaces/mcm/policies/policy-ma",
              "uid": "3f218d45-a7cc-11e9-8a41-005056a061f1",
            },
            "spec": Object {
              "namespaces": Object {},
              "policy-templates": Array [
                Object {
                  "objectDefinition": Object {
                    "apiVersion": "policies.ibm.com/v1alpha1",
                    "kind": "MutationPolicy",
                    "metadata": Object {
                      "label": Object {
                        "category": "System-Integrity",
                      },
                      "name": "mutation-policy-example",
                    },
                    "spec": Object {
                      "conditions": Object {
                        "ownership": Array [
                          "ReplicaSet",
                          "Deployment",
                          "DeamonSet",
                          "ReplicationController",
                        ],
                      },
                      "namespaceSelector": Object {
                        "exclude": Array [
                          "kube-system",
                        ],
                        "include": Array [
                          "default",
                          "kube-*",
                        ],
                      },
                      "remediationAction": "enforce",
                    },
                  },
                  "status": Object {
                    "Validity": Object {},
                  },
                },
              ],
              "remediationAction": "enforce",
            },
            "status": Object {
              "placementBindings": Array [
                "binding-ma",
              ],
              "placementPolicies": Array [
                "placement-ma",
              ],
              "status": Object {
                "cluster1": Object {
                  "aggregatePoliciesStatus": Object {
                    "policy-ma": Object {
                      "compliant": "NonCompliant",
                    },
                  },
                  "clustername": "cluster1",
                  "compliant": "NonCompliant",
                },
                "clusterhub": Object {
                  "aggregatePoliciesStatus": Object {
                    "policy-ma": Object {
                      "compliant": "Compliant",
                    },
                  },
                  "clustername": "clusterhub",
                  "compliant": "Compliant",
                },
              },
            },
          },
          "remediation": "enforce",
        },
        Object {
          "__typename": "Compliance",
          "clusterCompliant": "1/2",
          "metadata": Object {
            "__typename": "Metadata",
            "annotations": Object {
              "policy.mcm.ibm.com/categories": "SystemAndCommunicationsProtections,SystemAndInformationIntegrity",
              "policy.mcm.ibm.com/controls": "MutationAdvisor,CertificateManager",
              "policy.mcm.ibm.com/standards": "NIST,HIPAA",
              "seed-generation": "1",
            },
            "name": "policy-pod",
            "namespace": "mcm",
            "resourceVersion": "1257026",
            "selfLink": "/apis/policy.mcm.ibm.com/v1alpha1/namespaces/mcm/policies/policy-pod",
          },
          "name": "policy-pod",
          "namespace": "mcm",
          "placementBindings": Array [
            Object {
              "__typename": "PlacementBinding",
              "metadata": Object {
                "__typename": "Metadata",
                "name": "binding-pod",
                "selfLink": "/apis/mcm.ibm.com/v1alpha1/namespaces/mcm/placementbindings/binding-pod",
              },
            },
          ],
          "placementPolicies": Array [
            Object {
              "__typename": "PlacementPolicy",
              "metadata": Object {
                "__typename": "Metadata",
                "name": "placement-pod",
                "selfLink": "/apis/mcm.ibm.com/v1alpha1/namespaces/mcm/placementpolicies/placement-pod",
              },
            },
          ],
          "policyCompliant": "1/2",
          "raw": Object {
            "apiVersion": "policy.mcm.ibm.com/v1alpha1",
            "kind": "Policy",
            "metadata": Object {
              "annotations": Object {
                "policy.mcm.ibm.com/categories": "SystemAndCommunicationsProtections,SystemAndInformationIntegrity",
                "policy.mcm.ibm.com/controls": "MutationAdvisor,CertificateManager",
                "policy.mcm.ibm.com/standards": "NIST,HIPAA",
                "seed-generation": "1",
              },
              "creationTimestamp": "2019-07-16T13:16:46Z",
              "finalizers": Array [
                "propagator.finalizer.mcm.ibm.com",
              ],
              "generation": 19,
              "name": "policy-pod",
              "namespace": "mcm",
              "resourceVersion": "1257026",
              "selfLink": "/apis/policy.mcm.ibm.com/v1alpha1/namespaces/mcm/policies/policy-pod",
              "uid": "f61a25eb-a7cb-11e9-8a41-005056a061f1",
            },
            "spec": Object {
              "complianceType": "musthave",
              "namespaces": Object {
                "exclude": Array [
                  "kube*",
                ],
                "include": Array [
                  "default",
                ],
              },
              "object-templates": Array [
                Object {
                  "complianceType": "musthave",
                  "objectDefinition": Object {
                    "apiVersion": "v1",
                    "kind": "Pod",
                    "metadata": Object {
                      "name": "nginx-pod",
                    },
                    "spec": Object {
                      "containers": Array [
                        Object {
                          "image": "nginx:1.7.9",
                          "name": "nginx",
                          "ports": Array [
                            Object {
                              "containerPort": 80,
                            },
                          ],
                        },
                      ],
                    },
                  },
                  "status": Object {
                    "Validity": Object {},
                  },
                },
              ],
              "remediationAction": "inform",
            },
            "status": Object {
              "placementBindings": Array [
                "binding-pod",
              ],
              "placementPolicies": Array [
                "placement-pod",
              ],
              "status": Object {
                "cluster1": Object {
                  "aggregatePoliciesStatus": Object {
                    "policy-pod": Object {
                      "compliant": "Compliant",
                    },
                  },
                  "clustername": "cluster1",
                  "compliant": "Compliant",
                },
                "clusterhub": Object {
                  "aggregatePoliciesStatus": Object {
                    "policy-pod": Object {
                      "compliant": "NonCompliant",
                    },
                  },
                  "clustername": "clusterhub",
                  "compliant": "NonCompliant",
                },
              },
            },
          },
          "remediation": "inform",
        },
        Object {
          "__typename": "Compliance",
          "clusterCompliant": "1/2",
          "metadata": Object {
            "__typename": "Metadata",
            "annotations": Object {
              "kubectl.kubernetes.io/last-applied-configuration": "{\\"apiVersion\\":\\"policy.mcm.ibm.com/v1alpha1\\",\\"kind\\":\\"Policy\\",\\"metadata\\":{\\"annotations\\":{\\"policy.mcm.ibm.com/categories\\":\\"SystemAndCommunicationsProtections,SystemAndInformationIntegrity\\",\\"policy.mcm.ibm.com/controls\\":\\"MutationAdvisor,VA\\",\\"policy.mcm.ibm.com/standards\\":\\"NIST,HIPAA\\"},\\"name\\":\\"policy-role\\",\\"namespace\\":\\"mcm\\"},\\"spec\\":{\\"namespaces\\":{\\"exclude\\":[\\"kube*\\"],\\"include\\":[\\"default\\"]},\\"remediationAction\\":\\"inform\\",\\"role-templates\\":[{\\"apiVersion\\":\\"roletemplate.mcm.ibm.com/v1alpha1\\",\\"complianceType\\":\\"musthave\\",\\"metadata\\":{\\"name\\":\\"operator-role-policy\\",\\"namespace\\":\\"\\"},\\"rules\\":[{\\"complianceType\\":\\"musthave\\",\\"policyRule\\":{\\"apiGroups\\":[\\"extensions\\",\\"apps\\"],\\"resources\\":[\\"deployments\\"],\\"verbs\\":[\\"get\\",\\"list\\",\\"watch\\",\\"create\\",\\"delete\\",\\"patch\\"]}}],\\"selector\\":{\\"matchLabels\\":{\\"dev\\":\\"true\\"}}}]}}
",
              "policy.mcm.ibm.com/categories": "SystemAndCommunicationsProtections,SystemAndInformationIntegrity",
              "policy.mcm.ibm.com/controls": "MutationAdvisor,VA",
              "policy.mcm.ibm.com/standards": "NIST,HIPAA",
              "seed-generation": "1",
            },
            "name": "policy-role",
            "namespace": "mcm",
            "resourceVersion": "1201118",
            "selfLink": "/apis/policy.mcm.ibm.com/v1alpha1/namespaces/mcm/policies/policy-role",
          },
          "name": "policy-role",
          "namespace": "mcm",
          "placementBindings": Array [
            Object {
              "__typename": "PlacementBinding",
              "metadata": Object {
                "__typename": "Metadata",
                "name": "binding-role",
                "selfLink": "/apis/mcm.ibm.com/v1alpha1/namespaces/mcm/placementbindings/binding-role",
              },
            },
          ],
          "placementPolicies": Array [
            Object {
              "__typename": "PlacementPolicy",
              "metadata": Object {
                "__typename": "Metadata",
                "name": "placement-role",
                "selfLink": "/apis/mcm.ibm.com/v1alpha1/namespaces/mcm/placementpolicies/placement-role",
              },
            },
          ],
          "policyCompliant": "1/2",
          "raw": Object {
            "apiVersion": "policy.mcm.ibm.com/v1alpha1",
            "kind": "Policy",
            "metadata": Object {
              "annotations": Object {
                "kubectl.kubernetes.io/last-applied-configuration": "{\\"apiVersion\\":\\"policy.mcm.ibm.com/v1alpha1\\",\\"kind\\":\\"Policy\\",\\"metadata\\":{\\"annotations\\":{\\"policy.mcm.ibm.com/categories\\":\\"SystemAndCommunicationsProtections,SystemAndInformationIntegrity\\",\\"policy.mcm.ibm.com/controls\\":\\"MutationAdvisor,VA\\",\\"policy.mcm.ibm.com/standards\\":\\"NIST,HIPAA\\"},\\"name\\":\\"policy-role\\",\\"namespace\\":\\"mcm\\"},\\"spec\\":{\\"namespaces\\":{\\"exclude\\":[\\"kube*\\"],\\"include\\":[\\"default\\"]},\\"remediationAction\\":\\"inform\\",\\"role-templates\\":[{\\"apiVersion\\":\\"roletemplate.mcm.ibm.com/v1alpha1\\",\\"complianceType\\":\\"musthave\\",\\"metadata\\":{\\"name\\":\\"operator-role-policy\\",\\"namespace\\":\\"\\"},\\"rules\\":[{\\"complianceType\\":\\"musthave\\",\\"policyRule\\":{\\"apiGroups\\":[\\"extensions\\",\\"apps\\"],\\"resources\\":[\\"deployments\\"],\\"verbs\\":[\\"get\\",\\"list\\",\\"watch\\",\\"create\\",\\"delete\\",\\"patch\\"]}}],\\"selector\\":{\\"matchLabels\\":{\\"dev\\":\\"true\\"}}}]}}
",
                "policy.mcm.ibm.com/categories": "SystemAndCommunicationsProtections,SystemAndInformationIntegrity",
                "policy.mcm.ibm.com/controls": "MutationAdvisor,VA",
                "policy.mcm.ibm.com/standards": "NIST,HIPAA",
                "seed-generation": "1",
              },
              "creationTimestamp": "2019-07-16T13:19:03Z",
              "finalizers": Array [
                "propagator.finalizer.mcm.ibm.com",
              ],
              "generation": 16,
              "name": "policy-role",
              "namespace": "mcm",
              "resourceVersion": "1201118",
              "selfLink": "/apis/policy.mcm.ibm.com/v1alpha1/namespaces/mcm/policies/policy-role",
              "uid": "47dbc091-a7cc-11e9-8a41-005056a061f1",
            },
            "spec": Object {
              "namespaces": Object {
                "exclude": Array [
                  "kube*",
                ],
                "include": Array [
                  "default",
                ],
              },
              "remediationAction": "inform",
              "role-templates": Array [
                Object {
                  "apiVersion": "roletemplate.mcm.ibm.com/v1alpha1",
                  "complianceType": "musthave",
                  "metadata": Object {
                    "creationTimestamp": null,
                    "name": "operator-role-policy",
                  },
                  "rules": Array [
                    Object {
                      "complianceType": "musthave",
                      "policyRule": Object {
                        "apiGroups": Array [
                          "extensions",
                          "apps",
                        ],
                        "resources": Array [
                          "deployments",
                        ],
                        "verbs": Array [
                          "get",
                          "list",
                          "watch",
                          "create",
                          "delete",
                          "patch",
                        ],
                      },
                    },
                  ],
                  "selector": Object {
                    "matchLabels": Object {
                      "dev": "true",
                    },
                  },
                  "status": Object {
                    "Validity": Object {},
                  },
                },
              ],
            },
            "status": Object {
              "placementBindings": Array [
                "binding-role",
              ],
              "placementPolicies": Array [
                "placement-role",
              ],
              "status": Object {
                "cluster1": Object {
                  "aggregatePoliciesStatus": Object {
                    "policy-role": Object {
                      "compliant": "Compliant",
                    },
                  },
                  "clustername": "cluster1",
                  "compliant": "Compliant",
                },
                "clusterhub": Object {
                  "aggregatePoliciesStatus": Object {
                    "policy-role": Object {
                      "compliant": "NonCompliant",
                    },
                  },
                  "clustername": "clusterhub",
                  "compliant": "NonCompliant",
                },
              },
            },
          },
          "remediation": "inform",
        },
      ]
    }
    refreshControl={
      Object {
        "refreshCookie": "policy-refresh-interval-cookie",
        "reloading": false,
        "timestamp": "Tue Jul 23 2019 01:13:42 GMT-0400 (Eastern Daylight Time)",
      }
    }
    secondaryHeaderProps={
      Object {
        "information": "grc.header.tooltip",
        "links": Array [
          Object {
            "id": "create-policy",
            "label": "button.create.policy",
            "url": "/multicloud/policies/create",
          },
        ],
        "tabs": Array [
          Object {
            "id": "grc-overview",
            "index": 0,
            "label": "tabs.grc.overview",
            "url": "/multicloud/policies",
          },
          Object {
            "id": "grc-all",
            "index": 1,
            "label": "tabs.grc.all",
            "url": "/multicloud/policies/all",
          },
          Object {
            "id": "grc-findings",
            "index": 2,
            "label": "tabs.grc.findings",
            "url": "/multicloud/policies/findings",
          },
        ],
        "title": "routes.grc",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "render": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "render": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
